gcc/rust/expand/rust-macro-builtins-utility.cc

```diff
@@ -226,6 +227,86 @@ MacroBuiltin::env_handler (location_t invoc_locus, AST::MacroInvocData &invoc,
   return AST::Fragment ({node}, std::move (tok));
 }
 
+/* Expand builtin macro option_env!(), which inspects an environment variable at
+   compile time. */
+tl::optional<AST::Fragment>
+MacroBuiltin::option_env_handler (location_t invoc_locus,
+				  AST::MacroInvocData &invoc,
+				  AST::InvocKind semicolon)
+{
+  auto invoc_token_tree = invoc.get_delim_tok_tree ();
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (lex);
+
+  auto last_token_id = macro_end_token (invoc_token_tree, parser);
+  std::unique_ptr<AST::LiteralExpr> lit_expr = nullptr;
+  bool has_error = false;
+
+  auto start = lex.get_offs ();
+  auto expanded_expr = try_expand_many_expr (parser, last_token_id,
+					     invoc.get_expander (), has_error);
+  auto end = lex.get_offs ();
+
+  auto tokens = lex.get_token_slice (start, end);
+
+  if (has_error)
+    return AST::Fragment::create_error ();
+
+  auto pending = check_for_eager_invocations (expanded_expr);
+  if (!pending.empty ())
+    return make_eager_builtin_invocation (BuiltinMacro::OptionEnv, invoc_locus,
+					  invoc_token_tree,
+					  std::move (pending));
+
+  if (expanded_expr.size () != 1)
+    {
+      rust_error_at (invoc_locus, "%<option_env!%> takes 1 argument");
+      return AST::Fragment::create_error ();
+    }
+
+  if (expanded_expr.size () > 0)
+    if (!(lit_expr
+	  = try_extract_string_literal_from_fragment (invoc_locus,
+						      expanded_expr[0])))
+      return AST::Fragment::create_error ();
+
+  parser.skip_token (last_token_id);
+
+  auto env_value = getenv (lit_expr->as_string ().c_str ());
+  AST::Builder b (invoc_locus);
+  std::string option = "Option";
+
+  if (env_value == nullptr)
+    {
+      std::string none = "None";
+      std::unique_ptr<AST::Expr> none_expr
+	= std::unique_ptr<AST::Expr> (new AST::PathInExpression (
+	  b.path_in_expression (std::vector<std::string> ({option, none}))));
```

Comment:
so, the issue with this is that this won't necessarily pick up the "right" `Option` type, especially at the moment since gccrs does not handle the Rust prelude. a way to counter this would be to use this path in expression instead:

```suggestion
	  b.path_in_expression ({"core", "option", "Option", "None"})));
```

but this has an effect on your testcases, as you now need to add two modules to each of them

Reply:
Sure, I didn't realize it was better to use the full path. 
(Also I didn't know {"a"} => vector<string>, super useful)